### ISSUE TITLE
Automatic insertion of missing using directive uses CRLF in LF file if the new directive is the first in the file

### ISSUE THREAD

**Issue by PathogenDavid** [dotnet/roslyn#62976]
**Version Used**: Visual Studio 2022 17.3.0 Preview 5.0 / Compiler version: '4.3.0-3.22365.10 (dd078d4e)'. Language version: 10.0.

Also reproduced in Visual Studio 2022 17.2.6 / Compiler version: '4.2.0-4.22281.5 (8d3180e5)'. Language version: 10.0.

In both cases this issue was reproduced with all default settings (other than enabling this feature in step 2 below.)

**Steps to Reproduce**:

1. Clone this repository: https://github.com/PathogenPlayground/MissingUsingNewlineIssue
2. Under Tools > Options > Text Editor > C# > IntelliSense, ensure "Show items from unimported namespaces" is enabled.
3. Open the cloned solution
4. Open `Class1.cs`
5. Observe at the right-bottom of the editor that this file uses LF line endings.
6. Within `Class1`, start typing a new field `TheA`
7. IntelliSense will suggest `TheAType` from `TheANamespace`
8. Accept the suggestion
9. Note the line ending setting at the right-bottom of the editor

**Expected Behavior**:

The using directive is added with an LF line ending. The file still uses LF line endings.

**Actual Behavior**:

The using directive is added with a CRLF line ending. The file now uses mixed line endings.

----------

**Other info**:

This only happens if the automatically-added using directive will be the first in the file.

If at step 6/7 you use `TheIType` or `TheZType` (which will result in adding a using directive for `INamespace`/`ZNamespace` to the middle/end respectively), the issue does not occur.

**Comment by PathogenDavid** [comment]
Dug into this a bit, the carriage return is coming from here:

https://github.com/dotnet/roslyn/blob/0146540fc029667323d9aae46223b26e3ba2d04e/src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/LanguageServices/AddImports/AddImportHelpers.cs#L54-L58

This was originally added in https://github.com/dotnet/roslyn/commit/2033111f27e9d85cf073fdb66f1d03a2343d99c8 for https://github.com/dotnet/roslyn/pull/14674 to fix [this internal issue](https://devdiv.visualstudio.com/DevDiv/_workitems?id=226826&triage=true&_a=edit).

It's not clear why this newline is being added here. From what I can tell, the newline in the not-first-using-directive situation is added by [invoking the code formatter in `AbstractImportCompletionProvider`](https://github.com/dotnet/roslyn/blob/0146540fc029667323d9aae46223b26e3ba2d04e/src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/AbstractImportCompletionProvider.cs#L145) and this code is primarily dealing with leading trivia.

Maybe this was compensating for a limitation in the formatter which no longer exists? (I wanted to check, but those ancient versions of Roslyn can't be built anymore due to the shutdown of `dotnet.myget.org`.)

Removing that bit fixes the issue, and none of the `AddUsingTests`-related tests fail when doing so.

@CyrusNajmabadi I know it was a thousand years ago, but any idea what your intent was here? Maybe that internal issue says?

**Comment by CyrusNajmabadi** [comment]
elastic newlines should always be updated to match the newline file setting of that file.  So this would be a bug in the formatter.

**Comment by PathogenDavid** [comment]
Thanks, that's very helpful to know!

I did a bit more digging, here's what I found in case someone else wants to poke around (and so future me can remember.)

-----------

**Checking the first-line thing is important to the formatter bug**

At a formatter level, the issue is not specific to the first line. If you modify `AddImportHelpers.MoveTrivia` to *always* add an elastic CRLF to the new using, the new using will always result in mixed newlines.

---------------

**Looking for places handling elastic newlines**

I put a breakpoint in `GreenNode.HasAnnotation` to detect whenever something discovers an elastic newline is indeed elastic. Somewhat surprisingly I only found two places:

The first one is [something to do with determining the formatting anchor](https://github.com/dotnet/roslyn/blob/e541e3c7535be9c1e14a352e06d3859524c04473/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Formatting/Context/FormattingContext.cs#L422), I don't think it's relevant here.

The other looks more promising. Judging by the comment this is maybe intended to reformat the newlines:

https://github.com/dotnet/roslyn/blob/375af68543bd55d37dcd0773f8c0df6a92897aab/src/Workspaces/Core/Portable/CodeActions/CodeAction.cs#L340-L341

However, this code is also not called in the actual IDE case though. As noted in my previous comment, in the IDE case the reformatting happens in [this method](https://github.com/dotnet/roslyn/blob/0146540fc029667323d9aae46223b26e3ba2d04e/src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/AbstractImportCompletionProvider.cs#L145). I tried adding the "format any elastic whitespace" bit after that line, but sadly it doesn't seem like it's going to be that simple.

Either way, this isn't changing the elastic CRLF to an LF in the non-IDE case either. Need to dig into it further.

----------

**Unit test**

It took a hot second, but I did manage to get a unit test written for this. (Most of my investigation was done without this test, so I might've stopped minutes before a breakthrough.)

A lot of the test infrastructure around this feature really wants to be newline-ambivalent. As far as I can see there's no way to turn this off. I added a `TestParameters` option to control some of it, but the `TestWorkspace` deserialization still has some that I hard-disabled for now.

My test (as well as some scraps of fiddling around) can be found in [the roslyn62976 branch of my fork](https://github.com/PathogenDavid/roslyn/commit/bc2241907a2ce84fc2e6d0fd4e70cc756d2ef7d6).
